Enable the Permissions::query function to get the permission state of the Notifications API

RupinMittal opened the pull request:
#### 0e4995e19c781db613b6c88ec19dcdc118f8267f
<pre>
Enable the Permissions::query function to get the permission state of the Notifications API
<a href="https://bugs.webkit.org/show_bug.cgi?id=242818">https://bugs.webkit.org/show_bug.cgi?id=242818</a>

Reviewed by Sihui Liu and Chris Dumez.

Previously, Permissions::query would recieve a &quot;NotSupportedError&quot; when trying to query the
permission state of the Notifications API. Now the operation is supported and
Permissions::query will be able to call the necessary functions to obtain the permission
state of the Notifications API.

* LayoutTests/imported/w3c/web-platform-tests/permissions/all-permissions-expected.txt:
* LayoutTests/platform/ios-wk2/TestExpectations:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):

Canonical link: <a href="https://commits.webkit.org/252706@main">https://commits.webkit.org/252706@main</a>
</pre>

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -104,18 +108,48 @@ void WebPermissionController::tryProcessingRequests()
-                return;
 
+        // If the context is provided in the request, the API being queried is Notifications and so the permission
+        // will be obtained by using the context rather then sending a message to the UIProcess
```
This cannot be correct, I don't think any single WebContent process has a global view of notification permissions, only the UIProcess does. You need to IPC the UIProcess and get the permission from there.

szewai replied:
There is a Notification.permission, and it gets the state in web process

cdumez replied:
I just looked at the code and it is true that we currently sync the notification permissions from the UIProcess to the WebProcess and that Notification.permission() doesn't currently query the UIProcess. I think this was needed because the Web API is synchronous.

However, the new Web API for permissions is async. I'll defer to @youennf who worked on this more but I still feel that the permission API should get all its permissions from the UIProcess instead of having bypasses in the WebProcess for specification permission types.

Note that even if @youennf does decide that getting the permission from the WebProcess is OK, the current behavior of relying on the presence of a context or not to determine if this is Notification permission is ugly and definitely not the right design.

szewai replied:
The goal of this patch is to add support for querying notification's permission. According to spec, the query() result is the same as Notification.permission, so I think the idea to invoke existing functions in web process is not wrong.

And we don't cache permission result in UI process now. I think we may want to move notification permission state to UI process to be align with other features, but that's another task. 

I agree with you that the implementation is wrong to make the decision based on whether context is nullptr.

cdumez replied:
The UIProcess has the information via WebNotificationProvider::notificationPermissions().

cdumez replied:
Also, if we really want to do this in the WebProcess, we don't need the ScriptExecutionContext at all AFAICT, you can just get the permission from `WebProcess::singleton().supplement<WebNotificationManager>()`.

szewai replied:
Where is it in UIProcess? Do you mean in WebKit's code? For permission like geolocation, it's stored in WebKit's clients. 

Yes I think we can use WebNotificationManager.

cdumez replied:
I already said earlier but the UIProcess can get the information from `WebNotificationProvider::notificationPermissions()`. It will ask the client for the permissions.

szewai replied:
ohoh I thought by "UI process has the information" you mean WebKit in UI process caches the result. Yeah we may want to let web process pull the permission state from UI process instead of push the state to web process later.

RupinMittal replied:
I've changed the implementation here so that notifications is not treated as an edge case.

geoffreygaren commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -52,7 +52,11 @@ void WebPermissionController::query(WebCore::ClientOrigin&& origin, WebCore::Per
 
-    m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), WTFMove(completionHandler) });
+    if (descriptor.name == WebCore::PermissionName::Notifications)
+        m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), WTFMove(completionHandler), WeakPtr { context } });
```
If we're going to use the ScriptExecutionContext as a weak pointer, then it is best to declare a WeakPtr<ScriptExecutionContext>&& argument, rather than a ScriptExecutionContext* argument. Declaring what we need up front can make callers more efficient, and it also makes object lifetime more explicit.

RupinMittal replied:
This makes sense--the implementation for this pull request has now changed, but I'll keep this idea in mind for the future.

geoffreygaren commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -38,7 +38,7 @@ struct PermissionDescriptor;
 public:
     virtual ~PermissionController() = default;
-    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&) = 0;
+    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```
Apple and WebKit style put callbacks like CompletionHander in the last argument position. This makes reading the call sites easier, since you see all the arguments before the multi-line block of callback code.

So, you should put ScriptExecutionContext before CompletionHandler here.
```suggestion
    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```
```suggestion
    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```
```suggestion
    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```
```suggestion
    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```
```suggestion
    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```
```suggestion
    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```
```suggestion
    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```
```suggestion
    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```
```suggestion
    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```
```suggestion
    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&, ScriptExecutionContext*) = 0;
```

RupinMittal replied:
I see what you mean about the style here. Thank you for the feedback! The implementation for this pull request has now changed, but I'll keep this in mind for the future.

geoffreygaren commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -104,18 +108,48 @@ void WebPermissionController::tryProcessingRequests()
+                if (!takenRequest.context->isSecureContext())
+                    takenRequest.completionHandler(WebCore::PermissionState::Denied);
+                else {
+                    WebCore::NotificationPermission notiPer = client->checkPermission((takenRequest.context).get());
```
WebKit style uses whole words. I think by "notiPer" you meant "notificationPermission".

RupinMittal replied:
I see what you mean about the style here. Thank you for the feedback! The implementation for this pull request has now changed, but I'll keep this in mind for the future.

youennf commented on LayoutTests/http/tests/notifications/notification-permission-query.html:
```diff
@@ -0,0 +1,84 @@
+        navigator.permissions.query({ name: "notifications" }).then((status)=>{
+            Notification_permission = window.Notification.permission;
+            permissions_query = status.state;
+            shouldBeEqualToString("permissions_query", Notification_permission);
```
For other permissions, we prefer to not provide 'denied' state right away.
Instead, we prefer having prompt and when the page tries to request permission, we return deny for the request and we then expose denied as the permission.
This is a fingerprinting countermeasure that is done for geolocation and microphone/camera IIRC.
Can we apply it there as well?

szewai replied:
I am not sure I understand the change needs to be made. Are you suggesting the first query would return "prompt" by default until user/client explicitly denies the permission?

query() returns "prompt" by default and `testRunner.denyWebNotificationPermission(testURL);` does the denial process so it should return "denied" here.

szewai replied:
I see, WebNotificationManager::initialize actually initializes the permission states with WebProcessCreationParameters.
I think this is an existing issue. To fix this we just need to make the initialize() to ignore the denied permissions.

RupinMittal replied:
Yes, yt does make sense to implement fingerprinting countermeasures in a similar way to camera and microphone. For this pull request, the implementation has been changed to simply enable Permissions::query() to get the permission state of the Notifications API without receiving an error. I believe we've decided to deal with the fingerprinting issue in a later patch.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8850,6 +8850,12 @@ void WebPageProxy::queryPermission(const ClientOrigin& clientOrigin, const Permi
+    } else if (descriptor.name == PermissionName::Notifications) {
+#if ENABLE(NOTIFICATIONS)
+        name = "notifications"_s;
+        // Ensure that the true permission state of the Notification API is returned
```
WebKit comments should end with a period, per coding style.

RupinMittal replied:
The comment has been updated.

szewai reviewed (approved):
Looks good

RupinMittal committed:
Enable the Permissions::query function to get the permission state of the Notifications API
https://bugs.webkit.org/show_bug.cgi?id=242818

Reviewed by Sihui Liu and Chris Dumez.

Previously, Permissions::query would recieve a "NotSupportedError" when trying to query the
permission state of the Notifications API. Now the operation is supported and
Permissions::query will be able to call the necessary functions to obtain the permission
state of the Notifications API.

* LayoutTests/imported/w3c/web-platform-tests/permissions/all-permissions-expected.txt:
* LayoutTests/platform/ios-wk2/TestExpectations:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):

Canonical link: https://commits.webkit.org/252706@main
